tiatoolbox/dataloader/wsireader.py

```diff
@@ -0,0 +1,107 @@
+            output_dir: output directory to save the output, default=os.getcwd()/output
+            tile_objective_value: objective value at which tile is generated, default=20
+            tile_read_size_w: tile width, default=5000
+            tile_read_size_h: tile height, default=5000
+        """
+
+        self.input_dir = input_dir
+        self.file_name = os.path.basename(file_name)
+        if output_dir is not None:
+            self.output_dir = os.path.join(output_dir, self.file_name)
+
+        self.openslide_obj = openslide.OpenSlide(
+            filename=os.path.join(self.input_dir, self.file_name)
+        )
+        self.tile_objective_value = np.int(tile_objective_value)
+        self.tile_read_size = np.array([tile_read_size_w, tile_read_size_h])
+        self.objective_power = np.int(
+            self.openslide_obj.properties[openslide.PROPERTY_NAME_OBJECTIVE_POWER]
+        )
+        self.level_count = self.openslide_obj.level_count
+        self.level_dimensions = self.openslide_obj.level_dimensions
+        self.level_downsamples = self.openslide_obj.level_downsamples
+
+    def __exit__(self):
+        self.openslide_obj.close()
+
+    def slide_info(self, save_mode=True, output_dir=None, output_name=None):
+        """
+        WSI meta data reader
+        Args:
+            save_mode: save meta information as yaml file
+            output_dir: output directory to save the meta information
+            output_name: output file name
+
+        Returns:
+            displays or saves WSI meta information
+
+        """
+        input_dir = self.input_dir
+        if output_dir is None:
+            self.output_dir = output_dir
+        if output_name is None:
+            output_name = "param.yaml"
+        if self.objective_power == 0:
+            self.objective_power = np.int(
+                self.openslide_obj.properties[openslide.PROPERTY_NAME_OBJECTIVE_POWER]
+            )
+        objective_power = self.objective_power
+        slide_dimension = self.openslide_obj.level_dimensions[0]
+        tile_objective_value = self.tile_objective_value
+        rescale = np.int(objective_power / tile_objective_value)
+        filename = self.file_name
+        tile_read_size = self.tile_read_size
+        level_count = self.level_count
+        level_dimensions = self.level_dimensions
+        level_downsamples = self.level_downsamples
+
+        param = {
+            "input_dir": input_dir,
+            "output_dir": output_dir,
+            "objective_power": objective_power,
+            "slide_dimension": slide_dimension,
+            "rescale": rescale,
+            "tile_objective_value": tile_objective_value,
+            "filename": filename,
+            "tile_read_size": tile_read_size.tolist(),
+            "level_count": level_count,
+            "level_dimensions": level_dimensions,
+            "level_downsamples": level_downsamples,
+        }
+        if save_mode:
```

Comment:
It's giving both options, if the user wants to save it by default it's yml otherwise dict is returned which can be saved in the desired format. I think it would be best to provide default formats to make features compatible with each other. Let me know if you think otherwise.

Reply:
I think a better way to phrase my point is that if you add the functionality of creating a file on disk as part of this function then you will be duplicating code in every other function which does something similar. Returning a native dict object avoids this duplication and then the output can simply be easily passed into a serialise/dump/save function (e.g. in the CLI method) if you wanted to write to disk. i.e. I think teach function should do just one thing e.g. get the metadata or write a dict to disk but not both.

Reply:
I think a better way to phrase my point is that if you add the functionality of creating a file on disk as part of this function then you will be duplicating code in every other function which does something similar. Returning a native dict object avoids this duplication and then the output can simply be easily passed into a serialise/dump/save function (e.g. in the CLI method) if you wanted to write to disk. i.e. I think teach function should do just one thing e.g. get the metadata or write a dict to disk but not both.

Reply:
Updated as suggested in this commit https://github.com/TIA-Lab/tiatoolbox/commit/ff3073b1be351b2ea38649ea5d142823d551eb32 I haven't added utils.io but added save_yaml option in misc as dataloader will also be io.